### ISSUE TITLE
New rule: GREATEST/LEAST instead of CASE WHEN/IF when possible

### ISSUE THREAD

**Issue by iajoiner** [sqlfluff/sqlfluff#1798]
_Note: This is for general enhancements to the project. Please use the Bug report template instead to raise parsing/linting/syntax issues for existing supported dialects_
In all dialects that have `GREATEST/LEAST` code such as `CASE WHEN num_cats < 0 THEN 0 ELSE num_cats END` are bad practices because they are very long and need to be simplified as `GREATEST(num_cats, 0)` which is much shorter and a lot more readable.

**Comment by jpy-git** [comment]
This seems like a specific case of the logic described in #1797, would suggest also adding this case to L043 once the PR for that issue goes through.

**Comment by jpy-git** [comment]
@barrywhart @tunetheweb What's your opinion on making this a rule?

- It's pretty similar to L043 but doesn't use COALESCE so should be a separate new rule.
- `GREATEST`/`LEAST` are common but are not in the `ANSI` standard like `COALESCE` so we would need to filter out this fix for some dialects to avoid corrupting code. [This](https://stackoverflow.com/questions/3794451/greatest-and-least-in-sql-standard) SO post suggests it isn't available for T-SQL but may need to assess the other dialects as well.
- Otherwise looks fairly simple.

**Comment by tunetheweb** [comment]
I dunno. I can see the argument for it, but given it’s not in ANSI, I can also see the argument for writing more generic code and keeping code more portable. So seems quite an opinionated rule, that many might not agree with.

Perhaps it be an optional config to rule L043 instead? And rename it to “avoid unnecessary CASE statements”. You could check for `COALESCE` replacements or `GREATEST`/`LEAST` or both.

As a slight aside, I never liked rule L043 to be honest - sorry whoever added it - personally I find `COALESCE` a lot more difficult to “grok” and think using it for that use case (pardon the pun) is showing off you understand arcane syntax rather than useful to future developers but maybe I’m just grumpy that way (I find JavaScript developers like to show off how few lines of JS they can write too!). I’m less grumpy about `GREATEST`, `LEAST` as that IS more readable IMHO, but still have the concern about choosing “standard” SQL over specific where not too much extra effort.

**Comment by barrywhart** [comment]
I like it. I'm not too concerned about database portability, since in my experience, SQL is almost always written for one database _only_.

**Comment by tunetheweb** [comment]
Well I meant more programmer portability than database portability cause yes I agree code tends not to move. Then again that’s part of what a linter is for - to enforce best practices on those that might not know them. So maybe it is a good, stand-alone, rule for the dialects that support it.

**Comment by barrywhart** [comment]
Yes, I like it in the sense of helping less experienced SQL users. maybe analysts and data scientists.

About `COALESCE()`, we used it frequently at my last job doing data science projects. There may be a data person versus engineer person "divide" here. It may also depend on the situation. IIRC, we would often use it simply to ensure no `NULL` values, e.g. `COALESCE(col, "")`, similar to the old Python trick of `some_var or ""` to avoid `None` values. I could see in some cases wanting to use `CASE` to more explicitly "call out" what you're doing and why. 🤷‍♂️

**Comment by jpy-git** [comment]
So do we think make L043 more generic to “avoid unnecessary CASE statements” or a new rule for this? I'm easy either way :)

**Comment by tunetheweb** [comment]
Slight preference to expand L043. Seems similar enough to keep together.

But would also be happy with separate rule and maybe that reduces rule complexity and necessary config options.

**Comment by barrywhart** [comment]
Ditto.

If we make a separate rule, there's some risk we end up with 5 or 10 similar rules. Also from a branding point of view, having one rule that does all this makes SQLFluff seem a little smarter and more magical. 🎩🪄🐇

**Comment by jpy-git** [comment]
Nice, will pick this up tomorrow 😎
